ENH: optimize: ``trust-constr`` optimization algorithms [GSoC 2017]

antonior92 opened the pull request:
Hi All,

The thread in Pull Request #7729 became a little too large and GitHub is having a hard time loading the  page. Hence I decided to create this new pull request with the same content.

This commit contains two solvers I implemented for my GSoC project.:

1. An implementation of a Byrd-Omojokun Trust-Region SQP method  that solves 
equality constrained optimization problems of the type:

        minimize fun(x)
        subject to: constr(x) = 0

The underlying principles of this method are described in the following blog post ([here](https://antonior92.github.io/posts/2017/06/Byrd-Omojokun/)).

2. An implementation of a trust-region interior point method described
that solves general nonlinear programming problems:

        minimize f(x)
        subject to: constr_ineq(x) <= 0
                    constr_eq(x)    = 0

which uses the previously described SQP method as a substep. The underlying principle of this method can be found in another blog post ([here](https://antonior92.github.io/posts/2017/07/interior-point-method/)).

Both methods are based on the projected CG algorithm (described [here](https://antonior92.github.io/posts/2017/05/projected-CG/)) and are appropriate for large and sparse problems.

I tested both solvers on several problems from the CUTEst problem collection (the results are described [here](https://antonior92.github.io/posts/2017/00/NumericalResults/)). And while the performance (in number of iterations) is not as good as the one from the original paper yet, it came pretty close to it and could probably be improved with some fine tuning of the solver internal parameters.

antonior92 commented:
Me and @nmayorov are doing some final refactoring in the code and we expect the pull request to be ready for 1.1 scipy release. The main steps we have agreed upon are:
- [ ] Refactor new constraint classes;
- [ ] Improve solver return options;
- [ ] Improve documentation and write tutorial about the new method.

antonior92 commented on scipy/optimize/_constraints.py:
```diff
@@ -316,8 +299,15 @@ class BoxConstraint:
+    def __init__(self, lb=-np.inf, ub=np.inf, enforce_feasibility=False):
+        lb = np.asarray(lb)
+        ub = np.asarray(ub)
+        if np.all(lb == -np.inf):
```
Hi Nikolay,

I had time to take a look on your implementation. Really clean and nice.

Just about the future, it may worth considering saving ``ub`` and ``lb``. This way it will be easier to access them, for instance for using on finite differences.

This way, it may worth considering replacing this first few lines with:
```
self.lb = np.asarray(lb)
self.ub = np.asarray(ub)
```

nmayorov commented:
@pv @rgommers 

Need assistance. We decided that the old `bounds` format in `minimize` should be removed in the future and issue a `DeprecationWarning` [here](https://github.com/scipy/scipy/pull/8328/files#diff-98e852fc49f466bd2d642d5a610d28c0R579). Our substitute is `BoxConstraint(lb, ub)`, and when the old format is removed we can possible simplify it to `(lb, ub)` as in `least_squares`.

So the questions:

1. Does it sound right to you?
2. What should we write in `DeprecationWarning`? Like in which versions we want to remove it.
3. Meanwhile, should we just fix all tests to use the new format and not to trigger this warning?

andyfaff commented:
> Need assistance. We decided that the old bounds format in minimize should be removed in the future and issue a DeprecationWarning here. Our substitute is BoxConstraint(lb, ub), and when the old format is removed we can possible simplify it to (lb, ub) as in least_squares.

This is something that's going to need wider discussion on scipy-dev, as the way of providing bounds has implications in several other functions/class in `optimize`, and a large amount of users code might have to change to accomodate this.

nmayorov commented:
Actually, now I'm thinking about it and I'm not sure if this deprecations is necessary. The only bonus is the possibility to remove this `BoxConstraint` class in the future. But again it's a confusing road: first remove the sequence format and then reintroduce it with the different convention. This is not bad, but maybe not worth it.

antonior92 commented:
I think it would be better to keep both modes, without the deprecation. This pull request already has many complicated features. We may consider the deprecation latter on, if it is the case.

antonior92 commented on scipy/optimize/_differentiable_functions.py:
```diff
@@ -407,27 +407,45 @@ class LinearVectorFunction(object):
         self.H = sps.csr_matrix((self.n, self.n))
 
+    def change_x(self, x):
+        if not np.array_equal(x, self.x):
```
Hi Nikolay,

I know this is quite minor, but I think in order to make the implementation more similar to `VectorFunction`. It would be better if we define:
```
def update_x(self, x):
    self.x = x
    self.f_updated = False
```

and:

```
def fun(self, x):
    if not np.array_equal(x, self.x):
        self.update(x)
...
def jac(self, x):
    if not np.array_equal(x, self.x):
        self.update(x)
...
def hess(self, x):
    if not np.array_equal(x, self.x):
        self.update(x)
...
```

antonior92 commented:
Hi @nmayorov,

I have of proposal of modification for the canonical constraints. It involves the definition of a base class for ``VectorFunction``, ``LinearVectorFunction`` and ``IdentityVectorFunction``. I will call
this base class ``VectorFunctionBase``:

```
class VectorFunctionBase:

    def __init__():
        ...

    def fun(self, x):  # Returns Vector
        ...

    def jac(self, x):  # Returns Matrix (sparse or dense)
        ...

    def hess(self, x): # Return Matrix (sparse or dense) or LinearOperator
        ...
```

We will create a class that is basically a concatenation of `VectorFunctionBase`:

```
class ConcatenationOfVectorFunctionBase:

    def __init__():
        ...

    def fun(self, x):  # Returns List of Vector
        ...

    def jac(self, x):  # Returns List of Matrix (sparse or dense)
        ...

    def hess(self, x): # Returnl List of Matrix (sparse or dense) or LinearOperator
        ...
```

And once the user have passed the constraints we would form three structures:

1. One object containg ``ConcatenationOfVectorFunctionBase``.
2. One object containing a list of ``enforce_feasibility`` information for each constraint
3. One object containing a list of ``kind`` for each constraint.

We then could follow the following sequence:

I. Convert all constraints to constraints of the type `` c_eq(x) == 0`` or ``c_ineq(x) >=0``. We would use a flag ``ineq_eq_flag`` to indicate the type of the constraint,  and have a new object  ``ConcatenationOfVectorFunctionBase`` for the modified constraints.

```
(concatOfVectorFunction, enforce_feasibility, kind) -> (newConcatOfVectorFunction, ineq_eq_flag)
```

II. Concatenate inequalities constraints and equality constraints. Obtaining ``ConcatenationOfVectorFunctionBase`` containing only two functions, one for equality and other for inequalities constraints.

```
(concatofVectorFunction, enforce_feasibility, kind) -> newConcatOfVectorFunctionWithOnly2Constr
```

``ConcatenationOfVectorFunctionBase`` would assume the role of our ``CanonicalConstraint``.

It has some advantages:
1. `` ConcatenationOfVectorFunctionBase`` have a more clear meaning than ``CanonicalConstraint``;
2. It would allow us to keep considering the possibility that a inequality "interval" constraint can be an equality constraint when lb[i] == ub[i], which, as I have said is usefull for compaibility with CUTEst and IPOPT interfaces.
3. It would probably make the code more simple and easier to read.


Just to make it clear why I think this class would simplify things up: its API would guarantee that all constraints are called at once. Hence, consider:

```
concatofVectorFunction -> newConcatOfVectorFunction
```

Two of the new constraints from ``newConcatOfVectorFunction`` may be a function of a single constraint from ``concatofVectorFunction`` without we need to call a constraint from ``concatofVectorFunction`` twice.

ilayn commented:
Maybe you can rebase the commits where you feel confident such that the main steps are easier to track?

nmayorov commented:
We have sort of a problem here: a "flickering" test fail on the travis OSX configuration, see [here](https://travis-ci.org/scipy/scipy/builds/337718008?utm_source=github_status&utm_medium=notification) for example. 

My guess on what is happening here is described in #8208 by @pv ---something in complicated linear algebra routines, which are used a lot in this solver, is not deterministic. I also noted the warning
```
RuntimeWarning: internal gelsd driver lwork query error, required iwork dimension not returned. This is likely the result of LAPACK bug 0038, fixed in LAPACK 3.2.2 (released July 21, 2010). Falling back to 'gelss' driver.
    overwrite_b=overwrite)
```
It only appears in this configuration and may indicate that the LAPACK version used there is different from other configurations.

antonior92 commented:
Hi Nikolay,

I think something in the line of #8208 is a good guess. Nevertheless, it is worth pointing out that ``scipy.linalg.lstsq``, which is generating the warning, is not used in the portion of the code we are getting the error (I think ``scipy.linalg.lstsq`` is used only in ``test_projections.py``).

andyfaff commented:
I'm just trying to refactor the scalar minimizers (F: R^n->R) in optimize into class based form. See #8414. As such, that's required creation of a [`Function` object](https://github.com/andyfaff/scipy/blob/OPTIMIZE/scipy/optimize/optimize.py#L3208), which I'm thinking will end up as part of a public interface.
There is obviously a lot of overlap here with `ScalarFunction` in `scipy/scipy/optimize/_differentiable_functions.py`.

I have a few comments:

1) Will `ScalarFunction` end up being made public? If so, then we should probably find some way of merging them.
2) From a discussion with @jaimefrio, he pointed out it's a bad idea to save state (such as number of function evaluations) in such an object, because the same ScalarFunction may be reused across different minimizers, etc. The optimizer itself should be tracking the function evaluations.
3) All of the ScalarFunction parameters should probably be optional (and most of the __init__ could be streamlined). This is for two reasons, I think the design is nicer and it would enable easy inheritance of the object - for example users may be able to provide analytic func, grad, and hess, meaning they could easily override those methods (computation time may be saved if the user has fancy tricks).
4) Across Optimize grad and jacobian are used interchangeably, it would be good if the entire module settled on one. (Seems to be more heterogenous with the solvers).
5) I think the class based design in #8414, with the ability to iterate a single step at a time, is the way of the future, and will address many open issues. I've translated BFGS, LBFGSB and NelderMead so far, and it's a relatively simple transition - you just have to break out the iteration loop into a `__next__` method.
6) It would be good if a solid consensus was reached on how to describe upper/lower bounds, e.g. [(lower0, upper0)....] or np.r_[all_lower, all_upper].

Anyway, let me know what you think of the design I'm working on, and my comments, and let's see if we can converge somewhat.

antonior92 commented:
Hi @andyfaff,

Thank you a lot for your comments. I answer to them bellow:

1. In my opinion, we should keep ``ScalarFunction`` and ``VectorFunction`` as private classes. I don't know what @nmayorov thinks about this, but I think of these classes as a common implementation of features like finite-differentiation and quasi-Newton approximations. In the long term, I think it would be a easy way to make these features available to all the solvers. For now, it is only the cleanest way I found of making these feature available to the solver we are implementing.
2. I think if this class is not made public it should not be much of a problem, since the user cannot pass it twice to `minimize`. Furthermore, it can be hard to have a accurate count of function evaluations when using finite-differences approximations and quasi-Newton approximations.
3. Again, I think will be a problem only if we made the class public. 
4. I totally agree with you there. This is kind of messy right now.
5. I like the idea of introducing a little more structure to be shared across different solvers. I will study #8414 a little further and try to evaluate how we  would include these modifications.
6. I generally prefer ``np.r_[all_lower, all_upper]``. But now we allow both options in ``minimize`` using the class ``Bounds``. We considered deprecating ``[(lower0, upper0)....]``, but we though it would be better if we left it for latter on.

Anyway, I think it is important to create some structure for sharing features like finite-differences and hessian approximations for more than one solver and I think ``ScalarFunction`` and ``VectorFunction`` would be good for that, specially in the heterogeneous space of scipy solvers.
I don't think, however, our implementation is a good public interface.

About the common structure you are trying to give to all the other solvers, I generally think it to be a very good idea. But anyway, I don't think we should hold back this PR for too long and try to solve too many problems with it. Ideally I would like to try to keep the goal of merging this ``trust-const`` for scipy 1.1.

antonior92 commented:
> About the common structure you are trying to give to all the other solvers, I generally think it to be a very good idea. But anyway, I don't think we should hold back this PR for too long and try to solve too many problems with it. Ideally I would like to try to keep the goal of merging it for scipy 1.1.

Just adding something to this comment that I think I have not made entirely clear. While I think we should not try to solve too many problems with this PR, I also think it should try to give a sketch of the structures that latter on will be used to other solvers as well.

Hence, we need to better cordinate our work and try to reach some consensus on the general structure we will adopt in order to avoid implementing everything twice

nmayorov commented:
Hey Andrew,

Overall I think it's good to introduce some object oriented design to numerical optimization, and if scipy will be the first library to provide that --- it will be great. 

In fact I think @antonior92 made the first steps in this direction here by introducing constraint and `HessianUpdateStrategy` classes. I don't see a good reason not to do the same for the objective function. My feeling that these classes should be "thin" (just definition of an object) or interfaces to keep ability to use them with broad variety of solvers.

Also as numerical optimization algorithms operate iterative then adding the ability to advance them one step farther looks logical and doesn't contradict anything. But adding this ability to already implemented solvers in other languages might be difficult.

-----------------

For this pull request we decided with Antonio that we should aim to finish it for the current `minimize` framework. And we want to merge it within 2 weeks. After that we are both interested to discuss/develop a new framework you are thinking about and working at the moment.

nmayorov commented on scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py:
```diff
@@ -0,0 +1,521 @@
+        `constraints`. When bound constraints are present one extra element
+        will be include (in the end of the list) to account to the values
+        of these constraints Jacobian matrices.
+    v : list of ndarray
```
I think it needs some clarification for the case of interval constraint.

It seems to me that `v < 0` indicates that the upper bound is active and `v > 0` that lower bound is active. Am I correct here? It should be included into the doc.

Considering this I believe it makes sense to include `Bounds`, to indicate which is active and which is not. But perhaps a clearer approach is to include them separately (well, maybe not).

antonior92 replied:
Actually is the oposite :) ``v > 0`` for an active upper bound and ``v<0`` for an active lower bound. I just added a note about it.

Besides that, I think it is better to keep it as it is, without separating them.

nmayorov replied:
I see now, there are several places where the sign changes compared to "Numerical Optimization".

nmayorov commented:
I'm a little bit concerned that Returns displayed before Options in html. One way I was able to solve it, is to name Options as Parameters. The only difference is that it doesn't make this note
```
See also
For documentation for the rest of the parameters, see scipy.optimize.minimize
```
Should we use `Parameters`?

andyfaff commented on scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py:
```diff
@@ -0,0 +1,521 @@
+        Gradient of the objective function at the solution.
+    lagrangian_grad : ndarray, shape (n,)
+        Gradient of the Lagrangian function at the solution.
+    nfev : integer
```
Judging from past experience the user would like to know the total number of function evaluations, including those used in numerical calculation of grad and hess.

antonior92 replied:
A few commits ago I was using conceptual nfev/ngev/nhess... I changed it because I think  it make more sense to give the real number of function/gradient/hessian evaluations. It does have a clear meaning, it is easier to document, and I think it is more useful for the user.

Besides that, I think the number of iterations ``niter``, gives some information about the algorithm progress, and makes "conceptual function evaluations" more or less redundant.

andyfaff commented:
I really like ScalarFunction in `_differentiable_functions.py`. It's telling that I've come to a very similar design to the [Function](https://github.com/scipy/scipy/pull/8414/files#diff-c1c2c73ad9e5ddef5673e2e95d19020fR3210) class in #8414.
I have a quick question - how much computation time/function evaluations are saved by the memoisation of the immediate last `x`? Under what circumstances (e.g. during hessian calculation) is the memoisation triggered?

antonior92 commented on scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py:
```diff
@@ -0,0 +1,521 @@
+        Total execution time.
+    trust_radius : float
+        Trust radius at the last iteration.
+    penalty : float
```
No. I will explain it somewhere.

antonior92 commented on scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py:
```diff
@@ -0,0 +1,521 @@
+    penalty : float
+        Penalty function at last iteration.
+    tolerance : float
+        Tolerance for barrier subproblem at the last iteration.
```
Ok, I changed the documentation to "Only for problems with inequality constraints.". Much better indeed.

antonior92 commented on scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py:
```diff
@@ -0,0 +1,521 @@
+        Termination message.
+    method : {'equality_constrained_sqp', 'tr_interior_point'}
+        Optimization method used.
+    constr_violation : float
```
Ok, I tried to both order fields by importance while trying to group them by similarity. Take a look to see what you think.

antonior92 commented on scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py:
```diff
@@ -0,0 +1,521 @@
+    constr : list of ndarray
+        List of constraint values at the solution. This list will give
+        the values in the same order used  in `constraints`. When bound
+        constraints are present one extra element will be include(in
```
`I have struggled with this already.

The point is that ``v`` will have to include an element correponding to bounds. So I choose to keep ``constr``, ``jac``, ``v``, ``constr_nfev``, ``constr_njev``, and ``constr_nhev`` with the same length.

I am not entirely sure about it thought. Maybe we could include a single explanation "When bound constraints are present one extra element will be included (in the end of the list) to account to the values of these constraints." to all these structure.

nmayorov replied:
> I am not entirely sure about it thought. Maybe we could include a single explanation "When bound constraints are present one extra element will be included (in the end of the list) to account to the values of these constraints." to all these structure.

Sounds like a good idea.

antonior92 replied:
Do you know how to do it nicely? I tried it, but it just messed things up. I never know exactly what to do for making scipy documentation look nice.

antonior92 commented:
> Should we use Parameters?

I think so

antonior92 commented:
> I have a quick question - how much computation time/function evaluations are saved by the memoisation of the immediate last x?

The main reason for using memoization, is to avoiding calling the function one extra time for each gradient evaluation when using finite_diferences. So, assume that we are using finite diferences and call the gradient 300 time, we are avoiding calling the objective function 300 hundred times. Same thing when estimating the hessian using finite differences.

It would be possible to use memoization only for finite differences without losing performance, but I think it would only make the code more complicated.

andyfaff commented:
>  avoiding calling the function one extra time for each gradient evaluation when using finite_diferences

Ok, good to know. There is probably a much simpler way to achieve this, which would reduce the complicated nature of your implementation. 

`optimize._approx_fprime_helper`  provides an `f0` to the gradient calculation. I used that approach in the [Function class in #8414](https://github.com/scipy/scipy/pull/8414/files#diff-c1c2c73ad9e5ddef5673e2e95d19020fR3297).
I initially went through a lot of trouble to implement memoisation of the last provided `x`, and it was pointed out to me that it's a fragile way of doing things. So if you're only using memoisation to save one function calculation during a grad calculation (or one grad calculation during a hessian calculation), then you can stop using the memoisation, but not lose performance by using the approach in the PR link above. Start off with `func_and_grad`, and follow the trail from there.

Crucially, you can also test it. Copy the `Function` code from my PR into a notebook (plus any imports) and see how many function calls are required for a single `func_and_grad` calculation with finite differences (taking into account the `fd_method`), for a given number of parameters. Then compare it to your approach. If they're the same then the memoisation is not required.
If the finite difference method is forward differences (`fd_method='2-point'`), and there are N parameters for a scalar function. Theoretically the smallest number of function evaluations for `func_and_grad` you can get away with for a scalar function is `N + 1`. For '3-point' central differences it should be `2N + 1`. The testing for `Function` confirms that.

andyfaff commented on scipy/optimize/_constraints.py:
```diff
@@ -0,0 +1,285 @@
+        Method of computing the Jacobian matrix (an m-by-n matrix,
+        where element (i, j) is the partial derivative of f[i] with
+        respect to x[j]).  The keywords {'2-point', '3-point',
+        'cs'} select a finite difference scheme for the numerical estimation.
```
Perhaps you should permit absolute delta for the finite differences, as all the scalar minimizers do that.
(side note, I wish all the scalar minimizers would allow relative step, but that's another issue).

antonior92 replied:
I am using relative steps because it is how it is implemented in ``approx_derivative``, which I use internally. So it makes everything easier. And while I am not being consistent with scalar minimizers, I am consistent with Nikolay's ``least_squares``.

andyfaff commented on scipy/optimize/_constraints.py:
```diff
@@ -0,0 +1,285 @@
+    The scheme 'cs' is potentially the most accurate, but requires the function
+    to correctly handles complex inputs and be analytically continuable to the
+    complex plane. The scheme '3-point' is more accurate than '2-point' but
+    requires twice as much operations.
```
twice as many

antonior92 replied:
Ok

andyfaff commented on scipy/optimize/_constraints.py:
```diff
@@ -0,0 +1,285 @@
+    hess : {callable, '2-point', '3-point', 'cs', HessianUpdateStrategy, None}, optional
+        Method for computing the Hessian matrix. The keywords
+        {'2-point', '3-point', 'cs'} select a finite difference scheme for
+        numerical  estimation.  Alternativelly, objects implementing
```
Alternatively

antonior92 replied:
Ok

andyfaff commented on scipy/optimize/_constraints.py:
```diff
@@ -0,0 +1,285 @@
+
+
+class PreparedConstraint(object):
+    """Constraint prepared from a user defined constrained.
```
user defined constraint

antonior92 replied:
Ok

andyfaff commented on scipy/optimize/_numdiff.py:
```diff
@@ -177,8 +176,8 @@ def group_columns(A, order=0):
-                      bounds=(-np.inf, np.inf), sparsity=None, args=(),
-                      kwargs={}):
+                      bounds=(-np.inf, np.inf), sparsity=None,
+                      as_linear_operator=False, args=(), kwargs={}):
```
`as_linear_operator` should be last in the keyword arguments, to maintain the same order.

nmayorov replied:
I think it's better to order it like this, it is not a public function yet. But we should make it public.

antonior92 replied:
I think I agree with Nikolay here. We do not have to care much with compatibility, because the function has not been made public yet

andyfaff commented on scipy/optimize/_numdiff.py:
```diff
@@ -195,7 +194,7 @@ def approx_derivative(fun, x0, method='3-point', rel_step=None, f0=None,
-    method : {'3-point', '2-point'}, optional
+    method : {'3-point', '2-point', 'cs'}, optional
         Finite difference method to use:
             - '2-point' - use the fist order accuracy forward or backward
```
fist-->first

antonior92 commented:
Hi @andyfaff,

> optimize._approx_fprime_helper provides an f0 to the gradient calculation. I used that approach in the Function class in #8414.

Actually I think this would not work for ``trust_constr``:

- I do not have direct access to f0 (only to a barrier function) inside the method ``trust-constr``.
- I could oblige the algorithm to always call fun and grad together, but this would be ineficient, since I would call ``grad`` more times than I actually need to (About half of the time I call `fun`  I don't call grad afterwards).
- This would not be that simple. Because we need also to consider the case for which the hessian is being approximated by finite differences.

andyfaff commented:
Hmm, I'm wondering if there's an even better way.
`joblib.memory` implements a good memoization for numpy arrays. It's
probably a step to far at this stage to vendor that part of the code. It's
also possibly overkill because we just need to examine the x-array, rather
than having to deal with many different arguments.

Currently you're storing f, g, h for the last `x` value. How about writing
a really simple memoizer that caches many `x`-arrays, and stores f,g,h for
all of them? We could set an upper limit on the size of the cache to limit
memory effects.

From testing the time taken for `hash(x.tobytes)` is approximately similar
to `np.all(last_x == x)` (depending on size of `x`). The overhead from
retrieving other values is small. It should be possible to write a stripped
down memoizer like this quite simply. I suspect `hash` may not be quite
good enough though, might require something from hashlib.

andyfaff commented:
>
> From testing the time taken for `hash(x.tobytes)` is approximately similar
> to `np.all(last_x == x)` (depending on size of `x`).
>

Should've pointed out that the time test was for `x.size == 1000`, for
larger arrays working out hashes takes much longer. For `x.size==100` the
hashing approach is faster.
I would have thought for optimization x.size is typically going to be <
1000.
Are there any advantages of storing more than the last x value?

andyfaff commented:
<would need to add max cache size>

```
class Cache(object):
    def __init__(self, function, args=(), kwargs=None):
        self.kwargs = {}
        self.args = ()

        if kwargs is not None:
            self.kwargs.update(kwargs)

        self.x = {}
        self.function = function

    def __call__(self, x):
        h = hash(x.tobytes())
        if h in self.x:
            return self.x[h]
        else:
            v = self.function(x, *self.args, **self.kwargs)

            self.x[h] = v
            return v
```

nmayorov commented:
Hey Antônio,

I think I improved the Return documentation quite a bit. Also made it shorter. Some points to discuss:

1. I would like you to improve the documentation of `initial_trust_radius`,  `initial_barrier_parameter`, `initial_tolerance` to match approximately `initial_penalty`. A user at least should approximately understand what it is. Also include the reference to [1].
1. I suggest to make couple of renames: `tolerance -> barrier_tolerance`, `penalty -> constraints_penalty`, `trust_radius -> tr_radius`. What do you think?
1. What is the meaning of providing gradient of the objective function and Jacobians of the constraints? Why not Hessians as well?

antonior92 commented:
> Are there any advantages of storing more than the last x value?

I don't think so. 

Nevertheless, I agree with you that there is room for improvement. As pointed out by Nikolay sometime ago using ``not np.array_equal(x, self.x)`` is unefficient, since it always compare all elements when it need to compare only till the first difference is found.

> Would like you to improve the documentation of initial_trust_radius, initial_barrier_parameter, initial_tolerance to match approximately initial_penalty. A user at least should approximately understand what it is. Also include references to [1].

Ok, I can try to do it if you want me to.

> I suggest to make couple of renames: tolerance -> barrier_tolerance, penalty -> constraints_penalty, trust_radius -> tr_radius. What do you think?

it is ok for me.

> What is the meaning of providing gradient of the objective function and Jacobians of the constraints? 

Jacobian and gradients are usefull when checking the stationarity of the point. They are needed to see if you are in a stationary point regarding some constraint or regarding the objective function.

They are also needed for computing the lagrangian gradient, so I have to keep them internally anyway.

> Why not Hessians as well?

i) It is too much trouble to return the Hessian (at least at the time I was implementing it I tought so, maybe I was wrong); 2) the Hessian is often given by a structure that is not very useful for the user (e.g. linear operator).

nmayorov commented:
> As pointed out by Nikolay sometime ago using not np.array_equal(x, self.x) is unefficient, since it always compare all elements when it need to compare only till the first difference is found.

To be honest I don't remember saying this 😄 . And I don't see it to be true.

> Ok, I can try to do it if you want me to.

Basically because you understand it better.

antonior92 commented:
> To be honest I don't remember saying this smile . 

hahahaha really? I thought you have said it.

> And I don't see it to be true.

It make sense actually. A function ``np.array_not_equal`` could be more efficiently than using ``not np.array_equal``. Since ``np.array_equal`` always have to check the entire array.

> Basically because you understand it better.

Ok I will do it.

antonior92 commented:
> No wait, you confuse me, they should be equivalent. Imagine k-th elements differ, then array_equal will return False on this k-th element, whereas array_not_equal will return True on the same k-th element. So k comparisons in total. Right?

Be ``k`` the first element they differ and ``N`` the total aray size. For ``np.array_equal`` we need ``N`` comparisons always, because we have to check all the elments to stabilish the two arrays are equal.

For ``np.array_not_equal`` if a single element differ they will not be equal. So we have to check only k elements.

andyfaff commented:
I looked into a faster array comparison. It turns out that `np.all(a==b)`
is the quickest you can do it. Its because the comparison is vectorised in
numpy.

larsoner commented:
Probably need to wait for something like this at the NumPy end:

https://github.com/numpy/numpy/pull/8528

But if the problem is specific enough and the expected benefits large enough, a tiny short-circuiting Cython comparison function could be used instead.

nmayorov commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
+
+When :math:`\text{n-lb}_j = \text{n-ub}_j`, the method reads the constraint as an
+equality constraint and deals with it accordingly. Besides that, one-sided constraint
+may be specified by setting :math:`\text{n-lb}_j` or :math:`\text{n-lb}_j` to ``np.inf`` with
```
I think it should be "can", not "may" in all similar cases. Don't want to go into details (and I'm not sure), perhaps someone with a better knowledge of English can confirm.

antonior92 replied:
It is ok for me. I will change everything to "can"

nmayorov commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
-    ...          'jac' : lambda x: np.array([0.0, 1.0])})
+    >>> nonl_cons = NonlinearConstraint(cons_f, -np.inf, 1, jac=cons_J, hess='2-point')
 
+If the constraint first derivatives are also not known. They may be approximated
```
`The Jacobian of the constraints can be approximated by finite differences as well:`

antonior92 replied:
Ok

nmayorov commented:
There are "failed examples" from the tutorial. I think you need to add `# may vary` before "expected" answers (where it happens).

Example
```
# may vary
`gtol` termination condition is satisfied.
Number of iterations: 27, function evaluations: 17, CG iterations: 15, optimality: 8.85e-10, constraint violation: 2.22e-16, execution time: 0.044 s.
```

antonior92 commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
+   :nowrap:
+
+     \begin{eqnarray*} \min_x & f(x) & \\
+          \text{subject to: } & ~~~\text{n-lb}_j  \leq c_j(x) \leq \text{n-ub}_j, & ~~~ j = 1,...,M\\
```
I think I prefer ``c^L, c^U`` for proper constraints, and ``x^L, x^U`` for bound constraints.

Just changed the names (to appear in my next commit)

antonior92 replied:
I removed the indexing as well.

```
\begin{eqnarray*} \min_x & f(x) & \\
          \text{subject to: } & ~~~ c^l  \leq c(x) \leq c^u, &\\
           &  x^l  \leq x \leq x^u. & \end{eqnarray*}
```

antonior92 commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
+may be specified by setting :math:`\text{n-lb}_j` or :math:`\text{n-lb}_j` to ``np.inf`` with
+the appropriate sign.
+
+The method switches between two implementations depending on the problem definition.
```
Ok! Thank you

antonior92 commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
+""""""""""""""""""""""""""""
 
-subject to an equality and an inequality constraints defined as:
+The bound constraints are defined using :func:`Bounds` object. For the example,
```
Ok

antonior92 commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
-subject to an equality and an inequality constraints defined as:
+The bound constraints are defined using :func:`Bounds` object. For the example,
+the bound constraints  :math:`0 \leq  x_0  \leq 1` and :math:`-0.5 \leq  x_1  \leq 2.0`
+may be defined as:
```
ok

antonior92 commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
+may be defined as:
+
+    >>> from scipy.optimize import Bounds
+    >>> bnds = Bounds([0, -0.5], [1.0, 2.0])
```
ok

antonior92 commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
+Defining Linear Constraints:
+""""""""""""""""""""""""""""
+
+The linear constraints are defined using :func:`LinearConstraint` object.
```
ok

antonior92 commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
+
+The linear constraints are defined using :func:`LinearConstraint` object.
+For the example, the linear constraints :math:`x_0 + 2 x_1 \leq 1`
+and :math:`2 x_0 + x_1 = 1` may be rewritten as:
```
ok

antonior92 commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
+and defined using:
 
+    >>> from scipy.optimize import LinearConstraint
+    >>> lin_cons = LinearConstraint([[1, 2], [2, 1]], [-np.inf, 1], [1, 1])
```
Ok. I changed it to ``linear_constraint``.

antonior92 commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
+    >>> lin_cons = LinearConstraint([[1, 2], [2, 1]], [-np.inf, 1], [1, 1])
 
-The objective function and its derivative are defined as follows.
+Defining Nonlinear Constraints:
```
ok

antonior92 commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
 
-Then constraints are defined as a sequence of dictionaries, with keys
-``type``, ``fun`` and ``jac``.
+Alternatively, the Hessian may also be approximated using finite-differences:
```
ok

antonior92 commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
+
+Solving the Optimization Problem:
+"""""""""""""""""""""""""""""""""
+Now with the constraints and objective functions defined, the optimization
```
ok

antonior92 commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
+    >>> print(res.x)
+    [0.41494432 0.17011137]
+
+Finally, this method is suitable for large-scale problems because it solves the trust-region subproblem
```
I think it would be better to create a separate section with a large-scale problem. I will try to write it immediately after the ``Solving the Optimization Problem`` section.

I will try to use a sparse Jacobian as well.

antonior92 replied:
Ok just changed my mind about that. I believe the best way would be to include some scripts on ``scipy-cookbook`` and point to there from the tutorial. I will write more about that in a comment
outside the review.

antonior92 commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
+    ...     return [[2*x[0], 1], [2*x[0], -1]]
+    >>> def cons_H(x, v):
+    ...     return v[0]*np.array([[2, 0], [0, 0]]) + v[1]*np.array([[2, 0], [0, 0]])
+    >>> nonl_cons = NonlinearConstraint(cons_f, -np.inf, 1, jac=cons_J, hess=cons_H)
```
I agree with you that we need to explain what is ``v`` better.

I think, however, explaining how to compute the Jacobian and Hessian may be out of the scope of the tutorial. If someone do not know what is Hessian and Jacobian I don't think they will be able to grasp it from the tutorial anyway.

nmayorov replied:
I think you are right. But would it be much trouble to write mathematical formulas for the Jacobian and the Hessian? In this case a reader will only need to understand how they were coded, not how they were computed (well, you got my idea).

I don't mind though if you think that it is unnecessary.

antonior92 replied:
Maybe we could just point to the wikipedia page constaining the definition.

nmayorov replied:
To clarify: I mean writing the Jacobian and the Hessian for this particular constraints.

antonior92 replied:
Ahh, ok! I will write it out.

For some reason I understood you wanted to write it out for the general case.

antonior92 replied:
It make sense better sense now.

antonior92 commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
-    ...     dfdx0 = sign*(-2*x[0] + 2*x[1] + 2)
-    ...     dfdx1 = sign*(2*x[0] - 4*x[1])
-    ...     return np.array([ dfdx0, dfdx1 ])
+Computing the second derivatives of the constraints may be too troublesome to implement
```
I rewrote it as:

``` When the evaluation of the linear combination of the constraint Hessians is difficult to implement or computationally infeasible, one may use :class:`HessianUpdateStrategy`. Currently available strategies are :class:`BFGS` and :class:`SR1`.
```

The constraint is a vector function that does not has a Hessian in the traditional sense. Individual elements do have Hessians and we are interested in the linear combination of these Hessians.

The remaining of the text has been modified and should help the reader to understand this section.

antonior92 replied:
I was using the word ``second derivative`` to avoid going into these kind of detail.

antonior92 commented on doc/source/tutorial/optimize.rst:
```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
-    [2. 1.]
+    [0.41494432 0.17011137]
+
+The first and second-order derivatives of the objective function may also be approximated.
```
ok

antonior92 commented:
Hi @nmayorov 

I think I have addressed most of your comments now. What is missing is to decide how to proceed regarding the last section.

I agree with you that it would be nice to have a more convincing example illustrating  the advantages of using conjugate gradient instead of direct matrix factorization. After thinking a little about it,
however, I think the scipy tutorial is not a good place to do this. It would only make things more confusing.

I like the approach you adopted for your least-square tutorial. You gave a simple example illustrating the API and included a link to scipy-cookbook with more complicated examples. What do you think about adopting a similar approach for ``trust-constr`` ? I think the tutorial as it is now gives a nice idea about how the API works. More complicated and convincing examples can be included in ``scipy-cookbook``.

antonior92 commented:
I think I would prefer to try to finalize and merge this PR first and write jupyter notebooks for ``scipy-cookbook`` after that. What do you think about it?

antonior92 commented:
By the way, I forgot to comment: I removed comments trying to explaing why the method is good for large scale problems. I thought they were not convincing without an large-scale example.

The text is more focused in explaining the API now. In latter ``scipy-cookbook`` examples we can include these explanations together with more convincing examples.

nmayorov commented:
Antônio, I looked through the updated version and it looks good.

> I think I would prefer to try to finalize and merge this PR first and write jupyter notebooks for scipy-cookbook after that. What do you think about it?

I support that, let's do that. We want it to be merged :)

What about your benchmarks? Can you run them one more time?

Any other specific points we need to address?

antonior92 commented:
Hi Nikolay,

I have just run the benchmarks. The good news is that the modifications we made reduced the overhead a little bit. The bad news is that the solver is not converging in some problems it was converging before (usually problems with inequality constraints). I will investigate the reason for that and give you a more detailed explanation today (regardless of fixing the problem or not).

So basically, what I think it needs to be done:
- [x] Run benchmarks for the case all derivatives are available;
- [ ] Fix sover in order to converge in benchmarks it was congerging before;
- [ ] Run benchmarks for quasi-Newton / finite-difference approximations (at least a few different configurations);
- [ ] Fix eventual problems;
- [ ] Rebase;
- [ ] Merge.

nmayorov commented:
> The bad news is that the solver is not converging in some problems it was converging before (usually problems with inequality constraints).

Oh, it doesn't sound too good. I hope you will be able to find the reason. Consider this as an exercise to work with git history. But seriously, if you need help --- I should help, because I'm at least partially  responsible for this regression.

antonior92 commented:
To be completely sincerely, I think it is more likely to be something I have done. Your modifications seems less likely to be able affect the convergence.

Anyway, I think it is important for me to dig into this first. If I having trouble I will ask for help :)

I will keep you updated about my progress.

nmayorov commented:
Hi Antônio!

Good job. Is anything left to do?

antonior92 commented:
Hi @nmayorov, with my last pull request I just finished fixing everything I wanted to.

I will print the results in a readable format, so I can show them to you (and to others). Them I will receive final comments and suggestions, rebase and then merge.

I am in doubt about how to proceed about the rebase: should we compress everyving to a single commit ``ENH: optimize: ``trust-constr`` optimization algorithms``? Or should we do something more elaborated with multiple stages?

rgommers commented:
CI is unhappy about the last commit.

Rebasing into a single commit seems a bit rigorous. I'd say at least committing major parts (e.g. solver1, solver1_tests, solver2, solver2_tests, tutorial, benchmarks) separately would be good. If you'd prefer to leave as is and merge, that would also not be prolematic imho.

antonior92 commented:
I've run the solver again on the problems mentioned in [this](https://antonior92.github.io/posts/2017/00/NumericalResults/) blog post. The results are summarized next:

antonior92 commented:
The table bellow describes the performance of the SQP solver on large-scale equality constrained problems. There are no changes when compared with the original results.

|   name   |  n  |  m  | nnz  |  niters  | f evals  | CG iters |   time   |   opt    |  c viol  |
|---------:|----:|----:|-----:|---------:|---------:|---------:|---------:|---------:|---------:|
|  HAGER2  |10001|5000 |15000 |    7     |    7     |    6     |   5.09   |1.6947e-08|1.8914e-12|
|  HAGER3  |10001|5000 |15000 |    7     |    7     |    6     |   5.21   |6.2766e-08|1.8369e-12|
| ORTHREGA |2053 |1024 | 7168 |    73    |   113    |   178    |  11.21   |1.5134e-08|2.0262e-15|
| ORTHREGC |1005 | 500 | 3500 |    42    |    66    |    81    |   0.83   |8.8626e-08|8.1001e-16|
| ORTHREGD |10003|5000 |25000 |    16    |    16    |    16    |  59.46   |6.5561e-08|7.1054e-13|
| DTOC1ND  |2998 |1996 |13972 |    17    |    15    |    28    |   1.32   |7.4098e-08|2.2138e-13|
|  DTOC2   |2998 |1996 | 7984 |    12    |    12    |   221    |   1.27   |3.8127e-08|6.5081e-12|
|  DTOC3   |4499 |2998 |10493 |    10    |    10    |    9     |   2.23   |1.1603e-08|1.7461e-15|
|  DTOC4   |14999|9998 |34993 |    7     |    7     |    6     |  18.51   |2.8176e-08|6.5436e-12|
|  DTOC5   |9999 |4999 |14997 |    9     |    9     |    8     |   8.14   |2.9281e-08|2.0709e-12|
|  DTOC6   |2001 |1000 | 3000 |    19    |    19    |    38    |   0.5    |2.8538e-08|8.0491e-16|
| EIGENA2  |2550 |1275 |125000|    5     |    5     |    4     |   0.36   |2.8422e-14|0.0000e+00|
| EIGENC2  | 462 | 231 | 9261 |    29    |    36    |   295    |   0.44   |7.0736e-08|4.4869e-16|
|  ARTIF   |1002 |1000 | 3000 |    11    |    11    |    0     |   0.16   |0.0000e+00|4.9495e-10|
| BRATU3D  |4913 |3375 |23625 |    6     |    6     |    0     |   6.63   |0.0000e+00|1.7761e-10|

antonior92 commented:
The interior point method (''tr_interior_point'') was also tested on a set of large-scale problems and the results are displayed on the table bellow.

|   name   |  n  |  m  | nnz  |  niters  | f evals  | CG iters |   time   |   opt    |  c viol  |
|---------:|----:|----:|-----:|---------:|---------:|---------:|---------:|---------:|---------:|
| CORKSCRW | 456 | 350 | 1050 |    54    |    48    |   157    |   0.45   |9.1948e-09|5.2552e-13|
| COSHFUN  | 61  | 20  | 118  |    33    |    31    |   242    |   0.16   |7.8148e-08|0.0000e+00|
| DIXCHLNV | 100 | 50  | 2550 |    34    |    25    |   249    |   0.27   |2.2829e-08|5.5511e-16|
|  HAGER4  |2001 |1000 | 3000 |    40    |    31    |   194    |   1.16   |6.7160e-08|1.9438e-12|
| HIMMELBK | 24  | 14  | 336  |    89    |    96    |    90    |   0.23   |3.1748e-09|6.5024e-09|
|  NGONE   | 100 |1273 | 4996 |   291    |   519    |   773    |  28.39   |9.5064e-08|3.6803e-21|
| OPTCNTRL | 32  | 20  |  70  |    64    |    55    |    71    |   0.28   |1.4211e-14|7.8486e-08|
| OPTMASS  |1210 |1005 | 3216 |   148    |   260    |   139    |   2.12   |2.5807e-08|1.4572e-16|
| ORTHREGF |1205 | 400 | 3200 |    37    |    30    |    45    |   1.03   |3.0486e-08|5.7454e-14|
| SVANBERG | 500 | 500 | 4500 |    34    |    27    |   1293   |   1.35   |2.3870e-08|0.0000e+00|
| READING1 | 202 | 100 | 400  |    37    |    52    |    30    |   0.2    |3.1349e-09|8.3219e-08|

antonior92 commented:
I have also tested "trust-constr" on many nonlinear small problems to test their robustness. The results are displayed on the table bellow. For problems with no inequalities constraints the equality_constrained_sqp method is used, otherwise the interior point method tr_interior_point is used. 
The results are slightly better than the original results (third table from the blog post), with the solver converging in fewer iterations in many problems.


|   name   |  n  |  m  | nnz  |  niters  | f evals  | CG iters |   time   |   opt    |  c viol  |
|---------:|----:|----:|-----:|---------:|---------:|---------:|---------:|---------:|---------:|
|   HS7    |  2  |  1  |  2   |    9     |    9     |    8     |   0.02   |1.7663e-15|2.6645e-14|
|   HS10   |  2  |  1  |  2   |    15    |    12    |    16    |   0.05   |1.2104e-08|0.0000e+00|
|   HS11   |  2  |  1  |  2   |    12    |    8     |    11    |   0.04   |5.9848e-09|0.0000e+00|
|   HS13   |  2  |  1  |  2   |    1     |    1     |    0     |   0.0    |1.0000e-11|0.0000e+00|
|   HS14   |  2  |  2  |  4   |    10    |    7     |    6     |   0.04   |2.5894e-09|2.2204e-16|
|   HS16   |  2  |  2  |  4   |    13    |    10    |    13    |   0.05   |1.6644e-08|0.0000e+00|
|   HS17   |  2  |  2  |  4   |    25    |    21    |    31    |   0.09   |2.5942e-08|0.0000e+00|
|   HS19   |  2  |  2  |  4   |    36    |    34    |    36    |   0.13   |1.1693e-09|0.0000e+00|
|   HS20   |  2  |  3  |  6   |    10    |    8     |    9     |   0.04   |1.7323e-08|0.0000e+00|
|   HS22   |  2  |  2  |  4   |    14    |    10    |    10    |   0.04   |3.6337e-08|0.0000e+00|
|   HS24   |  2  |  3  |  6   |    11    |    7     |    6     |   0.03   |4.2946e-08|0.0000e+00|
|   HS26   |  3  |  1  |  3   |    22    |    22    |    41    |   0.06   |1.2174e-10|4.9481e-08|
|   HS28   |  3  |  1  |  3   |    3     |    3     |    3     |   0.01   |9.8334e-16|4.4409e-16|
|   HS31   |  3  |  1  |  2   |    12    |    8     |    18    |   0.05   |2.9458e-08|0.0000e+00|
|   HS32   |  3  |  2  |  6   |    15    |    10    |    12    |   0.05   |8.7322e-08|0.0000e+00|
|   HS33   |  3  |  2  |  6   |    17    |    16    |    17    |   0.06   |1.8812e-08|0.0000e+00|
|   HS39   |  4  |  2  |  6   |    16    |    16    |    18    |   0.04   |1.3941e-11|3.9848e-12|
|   HS46   |  5  |  2  |  6   |    29    |    49    |    70    |   0.08   |9.4806e-08|3.3307e-16|
|   HS51   |  5  |  3  |  7   |    3     |    3     |    3     |   0.01   |4.0993e-16|0.0000e+00|
|   HS52   |  5  |  3  |  7   |    3     |    3     |    3     |   0.01   |1.3767e-14|4.4409e-16|
|   HS53   |  5  |  3  |  7   |    17    |    9     |    9     |   0.06   |8.8678e-08|0.0000e+00|
|   HS63   |  3  |  2  |  6   |    13    |    8     |    7     |   0.04   |1.2935e-09|1.6548e-08|
|   HS64   |  3  |  1  |  3   |    34    |    30    |    53    |   0.11   |2.2107e-09|0.0000e+00|
|   HS65   |  3  |  1  |  3   |    22    |    17    |    32    |   0.09   |2.7051e-09|0.0000e+00|
|   HS70   |  4  |  1  |  2   |    46    |    48    |    87    |   0.19   |3.6633e-08|0.0000e+00|
|   HS71   |  4  |  2  |  8   |    17    |    13    |    21    |   0.08   |6.1559e-09|6.0423e-08|
|   HS72   |  4  |  2  |  8   |    26    |    21    |    41    |   0.1    |6.8264e-12|0.0000e+00|
|   HS73   |  4  |  3  |  12  |    18    |    14    |    23    |   0.07   |4.3877e-11|9.0206e-17|
|   HS74   |  4  |  5  |  12  |    18    |    12    |    11    |   0.08   |5.9282e-12|2.2775e-09|
|   HS75   |  4  |  5  |  12  |   102    |   100    |    99    |   0.51   |2.5834e-08|1.7468e-08|
|   HS77   |  5  |  2  |  6   |    18    |    20    |    34    |   0.05   |6.3707e-08|3.7303e-14|
|   HS78   |  5  |  3  |  11  |    5     |    5     |    4     |   0.01   |7.5462e-11|6.1410e-11|
|   HS79   |  5  |  3  |  8   |    7     |    7     |    12    |   0.02   |1.7792e-11|1.2566e-11|
|   HS80   |  5  |  3  |  11  |    17    |    10    |    8     |   0.06   |5.5397e-12|7.7953e-10|
|   HS81   |  5  |  3  |  11  |    17    |    10    |    8     |   0.06   |5.5391e-12|7.7954e-10|
|   HS83   |  5  |  3  |  13  |    42    |    39    |    73    |   0.21   |3.6121e-08|0.0000e+00|
|   HS84   |  5  |  3  |  15  |    12    |    11    |    24    |   0.06   |4.2925e-08|0.0000e+00|
|   HS85   |  5  | 21  | 102  |   133    |   237    |   136    |   0.65   |8.5643e-08|0.0000e+00|
|   HS86   |  5  | 10  |  37  |    16    |    12    |    29    |   0.05   |3.3449e-08|0.0000e+00|
|   HS93   |  6  |  2  |  12  |    22    |    16    |    55    |   0.09   |4.7860e-09|0.0000e+00|
|   HS95   |  6  |  4  |  20  |   167    |   165    |   168    |   0.45   |2.0541e-08|0.0000e+00|
|   HS96   |  6  |  4  |  20  |   213    |   219    |   219    |   0.62   |3.6676e-08|0.0000e+00|
|   HS97   |  6  |  4  |  20  |   539    |   537    |   582    |   1.65   |2.1298e-08|0.0000e+00|
|   HS98   |  6  |  4  |  20  |   385    |   393    |   563    |   1.2    |2.2296e-08|0.0000e+00|
|   HS99   |  7  |  2  |  14  |    86    |   149    |   158    |   0.18   |1.4264e-04|2.9104e-11|
|  HS100   |  7  |  4  |  19  |    20    |    16    |    62    |   0.07   |4.1090e-08|0.0000e+00|
|  HS104   |  8  |  5  |  17  |    43    |    36    |   138    |   0.23   |3.9791e-08|0.0000e+00|
|  HS105   |  8  |  1  |  2   |    56    |    51    |   137    |   0.31   |3.4738e-08|0.0000e+00|
|  HS106   |  8  |  6  |  17  |   261    |   288    |   319    |   0.88   |2.5370e-08|0.0000e+00|
|  HS107   |  9  |  6  |  34  |    12    |    10    |    19    |   0.05   |3.9453e-08|1.4591e-10|
|  HS108   |  9  | 13  |  39  |    57    |    64    |    91    |   0.19   |5.3699e-08|0.0000e+00|
|  HS109   |  9  | 10  |  42  |   1001   |   1000   |   1397   |   5.0    |2.5217e-05|4.4237e+04|
|  HS111   | 10  |  3  |  14  |    27    |    20    |    47    |   0.1    |3.4073e-08|8.4907e-12|
|  HS112   | 10  |  3  |  14  |    37    |    30    |    67    |   0.12   |1.2737e-08|5.5511e-17|
|  HS113   | 10  |  8  |  32  |    33    |    26    |    98    |   0.12   |7.7740e-12|0.0000e+00|
|  HS114   | 10  | 11  |  31  |    86    |    80    |   271    |   0.44   |5.3934e-08|4.5475e-13|
|  HS116   | 13  | 14  |  43  |   131    |   152    |   398    |   0.58   |4.5874e-08|0.0000e+00|
|  HS117   | 15  |  5  |  62  |    28    |    23    |    84    |   0.1    |1.8678e-08|0.0000e+00|
|  HS118   | 15  | 17  |  39  |    55    |    51    |   144    |   0.27   |5.3903e-08|0.0000e+00|
|  HS119   | 16  |  8  |  53  |    24    |    20    |    78    |   0.11   |1.7284e-08|5.5511e-16|

antonior92 commented:
Finally I also test other methods for approximating the Hessian and theirs convergence on the list of small problems I just mentiones. In the table bellow I show the number of iterations for the analitical hessian (``Analit``), and for ``SR1()``, ``BFGS(exception_strategy='skip_update')``, ``BFGS(exception_strategy='damp_update')`` and ``2-point`` approximations. When the solver fails to converge I give the number of iterations equals to ``x``.

|   name   |  Analit  |   SR1    |BFGS(skip)|BFGS(damp)| 2-point  |
|---------:|---------:|---------:|---------:|---------:|---------:|
|   HS7    |    9     |    9     |    9     |    9     |    9     |
|   HS10   |    15    |    36    |    36    |    36    |    15    |
|   HS11   |    12    |    12    |    17    |    17    |    12    |
|   HS13   |    1     |    1     |    1     |    1     |    1     |
|   HS14   |    10    |    10    |    10    |    10    |    10    |
|   HS16   |    13    |    12    |    15    |    14    |    13    |
|   HS17   |    25    |    28    |    41    |    53    |    25    |
|   HS19   |    36    |    37    |    39    |    38    |    36    |
|   HS20   |    10    |    18    |    15    |    17    |    10    |
|   HS22   |    14    |    14    |    14    |    14    |    14    |
|   HS24   |    11    |    18    |    24    |    29    |    11    |
|   HS26   |    22    |    32    |    x     |   676    |    22    |
|   HS28   |    3     |    51    |    52    |    52    |    3     |
|   HS31   |    12    |    12    |    24    |    30    |    12    |
|   HS32   |    15    |    17    |    18    |    15    |    15    |
|   HS33   |    17    |    22    |    62    |    49    |    17    |
|   HS39   |    16    |    21    |    18    |    18    |    16    |
|   HS46   |    29    |    31    |   415    |    50    |    29    |
|   HS51   |    3     |    19    |    19    |    19    |    3     |
|   HS52   |    3     |    21    |    23    |    23    |    4     |
|   HS53   |    17    |    28    |    29    |    29    |    17    |
|   HS63   |    13    |    14    |    83    |    30    |    13    |
|   HS64   |    34    |    43    |   535    |    47    |    34    |
|   HS65   |    22    |    20    |    28    |    33    |    22    |
|   HS70   |    46    |    72    |    40    |    43    |    38    |
|   HS71   |    17    |    20    |   161    |    31    |    17    |
|   HS72   |    26    |    x     |    x     |    x     |    26    |
|   HS73   |    18    |    18    |    21    |    21    |    18    |
|   HS74   |    18    |    24    |    27    |    37    |    18    |
|   HS75   |   102    |    95    |   115    |    22    |   102    |
|   HS77   |    18    |    26    |    24    |    20    |    18    |
|   HS78   |    5     |    8     |    10    |    10    |    5     |
|   HS79   |    7     |    11    |    67    |    15    |    7     |
|   HS80   |    17    |    17    |    16    |    15    |    17    |
|   HS81   |    17    |    16    |    16    |    30    |    17    |
|   HS83   |    42    |    30    |    27    |    27    |    43    |
|   HS84   |    12    |    13    |    x     |    17    |    12    |
|   HS85   |   133    |   121    |   173    |   150    |   133    |
|   HS86   |    16    |    24    |    26    |    31    |    16    |
|   HS93   |    22    |    42    |   236    |    x     |    22    |
|   HS95   |   167    |    43    |   279    |   225    |   179    |
|   HS96   |   213    |    99    |   112    |    59    |   208    |
|   HS97   |   539    |   606    |   991    |    x     |   620    |
|   HS98   |   385    |   396    |   409    |   900    |   425    |
|   HS99   |    x     |    x     |    x     |    x     |    x     |
|  HS100   |    20    |    22    |    33    |    33    |    20    |
|  HS104   |    43    |    78    |    92    |    92    |    43    |
|  HS105   |    56    |   493    |   700    |   762    |    64    |
|  HS106   |   261    |    x     |    x     |    x     |   243    |
|  HS107   |    12    |    17    |    22    |    31    |    12    |
|  HS108   |    57    |   188    |    x     |    x     |    57    |
|  HS109   |    x     |    x     |    x     |    x     |    x     |
|  HS111   |    27    |    x     |    x     |    x     |    27    |
|  HS112   |    37    |    42    |    43    |    52    |    37    |
|  HS113   |    33    |    24    |    41    |    41    |    27    |
|  HS114   |    86    |   110    |    94    |    74    |    93    |
|  HS116   |   131    |    x     |    x     |    x     |   111    |
|  HS117   |    28    |    72    |   505    |    x     |    33    |
|  HS118   |    55    |   892    |   892    |   892    |    55    |
|  HS119   |    24    |    56    |    48    |    51    |    24    |

antonior92 commented:
@mdhaber @rgommers @andyfaff @nmayorov @pv I would like to know if you guys have any final comments, doubts or suggestions?

This continuous integration problem is just because I introduced a warning and forgot to filter it. I will fix it soon. After that, if no one have any further modifications requests,  I intend to rebase (I will reduce this PR to something about 10-20 commits) and than merge.

antonior92 commented:
Hi @nmayorov, I just realized that there is no entry in the THANKS.txt file with your name.

Can I add a entry with your name so I can credit the implementation of ``trust-constr`` to you as well?
Something like:
```
Nikolay Mayorov for implementing least_squares function and trust-constr optimization method.
```
(and maybe add other contributions that I am not aware of)

mdhaber commented:
Results look great, Antonio. I wanted to test it out on a few problems but I don't want that to hold this up.

rgommers commented:
Rebased set of commits looks great Antonio.

Would you mind adding a short description under New features at https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.1.0?

nmayorov commented:
Hey Antônio! Don't bother with thanks for now. Great work finishing it! Add the description in the release notes and then let's merge it.

rgommers commented:
@nmayorov merge it now if you're happy! The longer it gets in master before branching the better. Release notes is a wiki entry, so not in this PR.

antonior92 commented:
I just wrote better messages for each commit. I think we are ready to merge now (once continuous integration finishes running)...

antonior92 commented:
By the way, the code for running the benchmarks:

https://gist.github.com/antonior92/3a4a62bd78a6cd60138cbfc549193e4c
https://gist.github.com/antonior92/a3cdb70dce21e10fc3874042bf0d2271

antonior92 commented:
I am getting some uncorrelated test errors from scipy.special.

It is strange because I was getting "all green" CI tests before and have not actually modified the code, only the commit messages.

rgommers commented:
> I am getting some uncorrelated test errors from scipy.special.

That's due to a change in numpy master, you can ignore that here. This is good to merge as is.

antonior92 commented:
Ok! I will merge it them. @nmayorov is that ok for you?

nmayorov commented:
Yes, go ahead!

rgommers commented:
Woohoo! Awesome job Antonio and Nikolay.
